Add code review etiquette document (#104)
Babylonpartners/ios-playbook: Etiquette/CODE_REVIEW.md

Code Review Etiquette
===========

> This document describes a general etiquete and code of conduct that every team member should follow, as well as some concrete guidelines and activities we use in the team related to code reviews.

Reviewer: ```suggestion
This document describes a general etiquette and code of conduct that every team member should follow, as well as some concrete guidelines and activities we use in the team related to code reviews.
```

> The purpose of these guidelines is to formalise some basic requirements for pull requests and receiving/providing feedback to improve the quality of code reviews and the team work in general.

Reviewer 2: ```suggestion
The purpose of these guidelines is to formalise some basic requirements for pull requests and receiving/providing feedback to improve the quality of code reviews and team work in general.
```
```suggestion
The purpose of these guidelines is to formalise some basic requirements for pull requests and receiving/providing feedback to improve the quality of code reviews and the team work in general.
```

Note: Guidelines in this document should be considered to be of equal importance

Note: This document is heavily inspired by [this blog post](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)

Everyone
--------

> * Be respecful to each other

Reviewer: ```suggestion
* Be respectful to each other
```

* Be humble. ("I'm not sure - let's look it up.")
* Don't use hyperbole. ("always", "never", "endlessly", "nothing")
* Don't use sarcasm or trolling

> * If you feel annoyed at something/someone don't make it affect your or others work. Talk to a friend, go for a walk, meditate, switch to something else.

Reviewer 2: This is advice for life, not just PRs ❤️

> * If you feel annoyed at something/someone don't make it affect your or others work. Talk to a friend, go for a walk, meditate, switch to something else.

Author: ```suggestion
* If you feel annoyed at something/someone, don't make it affect your or others' work. Talk to a friend, go for a walk, meditate, switch to something else.
```

* Talk synchronously, e.g. call or in person, prefer not to use chat for that, if there are too many "I didn't understand" or "Alternative solution:" comments. Post a follow-up
  comment summarizing the discussion.

> * Avoid selective ownership of code ("mine", "not mine", "yours"). 

Reviewer 2: ```suggestion
* Avoid selective ownership of code ("mine", "not mine", "yours"). Prefer referring to it as `our code`.
```

* Use "we" instead of "you", unless asking for opinion of specific individual ("What do you think ...?", but "Here we are doing ... What if we ...?",)
* Avoid using terms that could be seen as referring to personal traits. ("dumb",
  "stupid", "insane"). Assume everyone is intelligent and well-meaning.
* Be aware of negative bias with online communication. (If content is neutral, we assume the tone is negative.) Can you use positive language as opposed to neutral?

> * When using emojis be aware that their meaning can be different for different individuals - winking smile may not be always innosent

Reviewer: ```suggestion
* When using emojis be aware that their meaning can be different for different individuals - winking smile may not be always innocent
```

Hope this is correct (UK) English 🇬🇧

* Same with any other media format you decide to use

> * Do not expect your comments to be answered immediately. If you need an answer fast - talk in person (same applies to the Slack messages)

Reviewer 2: ```suggestion
* Do not expect your comments to be answered immediately. If you need an answer fast - talk in person (the same applies to the Slack messages)
```

> * It's fine to have disagreements and its hard to please everyone.

Reviewer: ```suggestion
* It's fine to have disagreements and it's hard to please everyone.
```

Creating Pull Request
--------

> 

Reviewer 3: Should we add a point suggesting to follow our branch naming conventions?

```suggestion

* When creating your branches, name them using a format like `<author>/<ticket>` or `<author>/<ticket>-<shortdescription>`
```

> * Carefully fill in pull request template:

Reviewer 2: ```suggestion
* Carefully fill in the pull request template:
```

> 	* title of the PR should follow format `[XXX-123] Description of the change` starting with a ticket number

Reviewer 2: ```suggestion
	* the title of the PR should follow the format `[XXX-123] Description of the change`, starting with a ticket number
```

> 	* explain the context and motivation, don’t assume familiarity with the history. Highlight how the goal was achived without going too much into details - people tend not to read long descriptions and if the long explanation is needed probably just pull request will be not enough (the change may need to go through proposal process). 

Reviewer: ```suggestion
	* explain the context and motivation, don’t assume familiarity with the history. Highlight how the goal was achieved without going too much into details - people tend not to read long descriptions and if the long explanation is needed probably just pull request will be not enough (the change may need to go through proposal process). 
```

> 	* don't strip out parts of the template unless it is absolutely unrelevant to the changes (i.e. change in the build script does not need any screenshots as it does not touch UI)

Reviewer: ```suggestion
	* don't strip out parts of the template unless it is absolutely irrelevant to the changes (i.e. change in the build script does not need any screenshots as it does not touch UI)
```

	* don't forget labels (it's better to set them after PR is created as github sometimes fails with 405 status code and we don't know why)

> * explicitly request review from or mention team mates you specifically want to involve in the discussion

Reviewer 3: ```suggestion
* Explicitly request review from or mention team mates you specifically want to involve in the discussion (in addition to the ones that will be assigned to it automatically by PullAssigners)
```

> * Self-review your changes, adding comments where you think you can get questions (it may worth adding these as code or documentation comments). You can use github suggestions and apply them right away

Reviewer 3: I agree with the idea (that GitHub suggestions should be used when possible), but don't really get what this is mentioned here in relation to this "self-review your code" point? At least I didn't understand what this had to do with anything when reading this point 😅 

It's rightfully already mentioned in the next section ("Review Pull Requests") and has its place down there, so I think adding it here too adds confusion more than anything.

```suggestion
* Self-review your changes, adding comments where you think you can get questions (it may worth adding these as code or documentation comments)
```

Author: I mean that if when doing self-review one spots small issues they can be addressed with GitHub suggestions.

Reviewer 3: ah got it, keep it then 👍

> * Keep pull requests small, ideally it shouldn't be more than **800 additions/deletions**

Reviewer 2: ```suggestion
* Keep your pull requests small, ideally they shouldn't be more than **800 additions/deletions**
```

Reviewer 2: Also is it 800 total or 800 additions?

Author: That's the point where I feel we don't have a common understanding and that's something that should be formalised in the #110 , not sure though what's the best place to settle it, this PR or the danger's one.

Reviewer 3: This is a very good question, something I wondered in the [Proposal for Danger](https://github.com/Babylonpartners/ios-playbook/pull/110/files#diff-7e81422226f62792b97195e580c16d7bR173) too 😉  we need to clarify that

Reviewer 4: I would add a line saying:
`If your PR has more than the limit, you need to provide a valid reason why it is so. For example: refactor is often "all or nothing" and cannot be split into smaller pieces.

* Create a draft PR for work that is in progress, but only do that if you seek for opinions on your work

> * Mark PR with "Blocked" label if it has any dependency that should stop it from being merged (it's fine for pull request be both "blocked" and "ready for review", but it shouldn't be "ready for review" and "wip")

Reviewer 2: ```suggestion
* Mark PR with the "Blocked" label if it has any dependency that should stop it from being merged (it's fine for pull request be both "blocked" and "ready for review", but it shouldn't be "ready for review" and "wip")
```

* Give your team mates a reasonable time to review your changes, if the change is critical use "Top priority" label and post it in the team channel

> * Do not leave stale pull requests, seek for help/actions or close them (make sure you don't delete your local branch if you may need it later)

Reviewer 2: ```suggestion
* Do not leave stale pull requests, seek for help/actions or close them (make sure you don't delete your local branch if you may need it later). If you find one that appears to have been for a week, please warn the author. At the two weeks mark it can be closed.
```

> * It's our common effort to push product forward and we all do equally valuable work. That in mind when creating PR try to review some PRs, avoid having more than X opened PRs

Author: we should indeed. @NSMyself did you have a chance to look at our stats? I'd say we could maybe put 3 here as we concluded 4 is probably too big considering the size of the team and 2 feels too restrictive as it does not take into account the complexity of PRs (there might be more really trivial PRs).

P.S. I think that's not something that should be strictly enforced by our tools btw.

> * It's our common effort to push product forward and we all do equally valuable work. That in mind when creating PR try to review some PRs, avoid having more than X opened PRs

Reviewer 5: Nope, sorry, I've been focusing on HM. No time right now

> * It's our common effort to push product forward and we all do equally valuable work. That in mind when creating PR try to review some PRs, avoid having more than X opened PRs

Reviewer 3: We might also reevaluate that value once we enforce smaller PRs and encourage splitting features into iterative PRs – like @Ben-Henshall suggested – since that might mean smaller / more trivial PRs... but more PRs per team member too?

Reviewing Pull Request
--------

* Accept that many programming decisions are opinions. Discuss tradeoffs, which
  you prefer, and reach a resolution quickly. There are always multiple ways of doing things, and they may be equally acceptable
* Focus not on finding flaws in the code (no one writes perfect code) but on understanding the change in the first place

> * Encourage things you like, try not only comment on what you don't like

Reviewer 3: 👍

* Before suggesting to change code think if it brings any real value or if it is just your personal preference. Is this code something you can live with or not?
* Do not try to make everything perfect (it never will be and otherwise we'll never deliver anything), but do not compromise on quality
* Do not comment if you don't have any suggestions to address your comment

> * Use github suggestions as much as possible

Reviewer 3: ```suggestion
* Use GitHub suggestions as much as possible
```

* If you feel strong against the change - request changes and explain why you are doing that. Do not request changes for typos, missing documentation comments or code style violations, trust your colleagues to address them.

> * Approve changes if you are confident

Reviewer 2: ```suggestion
* Approve changes if you are confident about the changes
```

* Comment on the changes if something is not clear to you and needs further clarification or actions
* If you are not familiar with the domain ask for more context (better in person) rather than skipping to something more familiar
* Ask, don't tell. ("What do you think about...?", "How about ...?", "Should we try ...?")

> * Be breif but clear in your comments (code review is not about proving who is more correct or knows more)

Reviewer: ```suggestion
* Be brief but clear in your comments (code review is not about proving who is more correct or knows more)
```

* Unless explicitly asked for it do not suggest code to be written in a specific way - let the author to find or suggest a solution themselves, they may come up with a better one
* Submit all the comments as a single review, even if there is just one comment
* If you see a stale pull request without any updates - ask for the status update in the comment

> * Do not be a blocker - if you requested changes and they were addressed make sure you review it

Author: > might consider it nagging

yes, they may =)

I'll check if there is a point (I think there is) about re-requesting reviews. The flow should be - move PR back to "Ready for review" and re-request reviews from those whose comments you've addressed or reach to them personally.

> * Do not merge others pull requests - they may be waiting for some one else review

Reviewer 3: ```suggestion
* Do not merge others pull requests - they may be waiting for someone else review
```

[ ❔] Given we use a merge bot, should this rule be re-worded accordingly? Also not sure this rule apply to us, as to my understanding our own etiquette says that we should not merge PRs ourselves, and add the "Merge" label when we are the 2nd approver of the PR and there is no pending discussion or doubt? (which is rightfully mentioned in "Responding to feedback" section below already btw 👍 )

Author: > My personal preference is that only the authors can merge

I agree with that (and so should change the line about adding "merge" label), but again it feels like there is no one common approach here, so having this written will help in settling that

Author: @AliSoftware the "Responding to feedback" section which is from the point of author of the changes, so that means that author should be the one adding this label (there always can be exception like if they are on holidays but change is good to go otherwise)

Reviewer 3: Ok. I still think that the implicit "Only merge your own PRs" here (as opposed to "don't merge other's PRs") might be confusing to new joiners (they might think of it actually meaning "hitting the Merge button on GitHub")

```suggestion
* Do not merge others pull requests - they may be waiting for someone else review. In general, the PR author should be the one adding the "Merge" label to merge their own PRs (except rare cases like if the author is on holidays but the PR is otherwise good to go)
```

Responding to feedback
-------------------------

> * Wait before answering to the single comments, reviewer may have more (though they should prefer to send all comments at once)

Reviewer 2: ```suggestion
* Wait before answering to a single comment, the reviewer may be still writing more of them (though they should prefer to send all comments at once)
```

* Be grateful for the reviewer's suggestions. ("Good call. I'll make that
  change.")

> * If you disagree with suggestion explain way and avoid template answers like "out of the scope", "thank you for your opinion". If you think that suggestion is a personal preference and does not bring value - say so

Reviewer 3: ```suggestion
* If you disagree with a suggestion, explain why and avoid template answers like "out of the scope" or "thank you for your opinion". If you think that suggestion is a personal preference and does not bring value - say so
```

* Assume the best intention from the reviewer's comments.
* If a reviewer seems aggressive or angry or otherwise personal, consider if it is intended to be read that way and ask the person for clarification of intent, in person if possible. Do not escalate online discussions.
* If you can't reach agreement seek for more opinions
* Don't be too defensive about your code, someone will rewrite it anyway sooner or later (it might be yourself as well)
* Try to answer all comments. You can review your own code for that instead of sending individual replies
* Add "work in progress" label while addressing review comments
* Re-request review when you addressed the comments (either implementing suggested change or replying to the comment) and label it with "Ready for review" again

> * Do not rebase and force-push in your pull request branch after you recieved any comments, use merge intsead. All pull requests are squashed when they are merged into the base branch so the history of your changes won't matter. It's fine to rebase and force-push if there are no comments

Reviewer: ```suggestion
* Do not rebase and force-push in your pull request branch after you received any comments, use merge instead. All pull requests are squashed when they are merged into the base branch so the history of your changes won't matter. It's fine to rebase and force-push if there are no comments
```

Reviewer 6: (not leaving a suggestion since it'd probably conflict with Inami's suggestion)

It would be nice if we could add a little bit on the end explaining the reasoning for this - that reviewers can view changes since they last reviewed to make re-reviewing easier

>This allows reviewers to view changes on the PR since their last review, rather than re-reviewing a lot of the code they have previously reviewed.

Author: 👍 

btw Github allows to edit others comments and so amend their suggested changes (a loot of roof for abuse there! 😈 ).

> * Do no merge pull request with pending requests for changes unless you agreed with their authors that it was addressed

Author: ```suggestion
* Do not merge pull request with pending requests for changes unless you agreed with their authors that it was addressed
```

> * When you have two approvals and don't wait for review from anyone else - put a "Merge" label for bot to merge your changes.

Author: ```suggestion
* When you have two approvals and have no pending questions or requests for changes and don't wait for review from anyone else - put a "Merge" label for bot to merge your changes.
```

PR Parties
--------

> To improve colloboration every week (on Monday) we take time to review pull request together in small groups, what we call "PR parties"

Reviewer: ```suggestion
To improve collaboration every week (on Monday) we take time to review pull request together in small groups, what we call "PR parties".
```

Reviewer 3: I first read that as "to improve collaboration every week, ..." so took me a second to get the meaning of this. Maybe rewriting it like this would help?
```suggestion
To improve collaboration, we take time to review pull request together in small groups every week (on Monday). This is what we call "PR parties".
```

- Each group should include at least one senior developer. 

> - Each group will have few pull requests assigned to them. 

Reviewer 3: ```suggestion
- Each group will have a few pull requests assigned to them
```

> - When done with assigned PRs group is free to review something else (i.e. each other pull requests) or finish the meeting

Reviewer 3: ```suggestion
- When done with their assigned PRs, the group is free to review something else (i.e. each other's pull requests) or finish the meeting
```

Pull Assigners
--------

> To balance the workload and avoid knowledge silos we use Pull Assigners to assign reviewers.

Reviewer 3: ```suggestion
To balance the workload and avoid knowledge silos we use [Pull Assigners](https://pullpanda.com/assigner) to assign reviewers.
```

> - Each pull request will be assigned three random reviewers from the team

Reviewer 2: ```suggestion
- Each pull request will be assigned to three random reviewers from the team
```

- Two approvals are enough to merge pull request (there shouldn't be pending requests for changes though)

> - Anyone else is still welcome to review any pull request and author can request for reviews from specific team mates

Reviewer 6: Something I think we should clarify is "protocol" for when your PR has two or more reviewers that are absent and cannot review the PR. I am unsure if this would be a technical solution (There's a small discussion about this in https://github.com/pullreminders/backlog/issues/158) or if it's an agreement that we re-request reviews if there are not 2 or more eligible reviewers requested.

Reviewer 6: I think this will need a wider discussion with the team, so does not have to. be changed in this PR

> - Anyone else is still welcome to review any pull request and author can request for reviews from specific team mates

Author: ```suggestion
- Anyone else is still encouraged to review any pull request and author can request reviews from specific teammates
```

Author: I mean that not being requested for review by PullAssigners shouldn't stop anyone from reviewing PRs. The bot is merely a helper here (to try to balance the workload), not an enforcer, at least ATM. I think if we see that this is not enough then we can think about enforcing reviews but then the issue of people being unavailable will have a bigger effect.

Reviewing a Pull Request
--------

* Accept that many programming decisions are opinions. Discuss tradeoffs, which
  you prefer, and reach a resolution quickly. There are always multiple ways of doing things, and they may be equally acceptable
* Focus not on finding flaws in the code (no one writes perfect code) but on understanding the change in the first place
* Encourage things you like, try not only comment on what you don't like
* Before suggesting to change code think if it brings any real value or if it is just your personal preference. Is this code something you can live with or not?
* Do not try to make everything perfect (it never will be and otherwise we'll never deliver anything), but do not compromise on quality
* Do not comment if you don't have any suggestions to address your comment
* Use GitHub suggestions as much as possible
* If you feel strong against the change - request changes and explain why you are doing that. Do not request changes for typos, missing documentation comments or code style violations, trust your colleagues to address them.

> * Approve changes if you are confident about them

Author: ```suggestion
* Approve changes if you are confident about them. 
* If you are the first reviewer to approve the changes add the label "Needs one reviewer" so that others can see which PRs already have a feedback and which don't have any yet.
```

* Do not be a blocker - if you requested changes and they were addressed make sure you review it
* Do not merge others pull requests - they may be waiting for some one else review

Responding to feedback
-------------------------

* Wait before answering to the single comments, reviewer may have more (though they should prefer to send all comments at once)
* Be grateful for the reviewer's suggestions. ("Good call. I'll make that
  change.")
* If you disagree with suggestion explain way and avoid template answers like "out of the scope", "thank you for your opinion". If you think that suggestion is a personal preference and does not bring value - say so
* Assume the best intention from the reviewer's comments.
* If a reviewer seems aggressive or angry or otherwise personal, consider if it is intended to be read that way and ask the person for clarification of intent, in person if possible. Do not escalate online discussions.
* If you can't reach agreement seek for more opinions
* Don't be too defensive about your code, someone will rewrite it anyway sooner or later (it might be yourself as well)

> * Try to answer all comments. You can review your own code for that instead of sending individual replies

Author: ```suggestion
* Try to answer all comments, don't make teammates feel like you are ignoring them. You can review your own code for that instead of sending individual replies so that all your answers are posted at once.
```

===========

This document describes a general etiquette and code of conduct that every team member should follow, as well as some concrete guidelines and activities we use in the team related to code reviews.
The purpose of these guidelines is to formalise some basic requirements for pull requests and receiving/providing feedback to improve the quality of code reviews and teamwork in general.

Note: Guidelines in this document should be considered to be of equal importance

Note: This document is heavily inspired by [this blog post](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/)

Everyone
--------

* Be respectful to each other
* Be humble. ("I'm not sure - let's look it up.")
* Don't use hyperbole. ("always", "never", "endlessly", "nothing")
* Don't use sarcasm or trolling
* If you feel annoyed at something/someone, don't make it affect your or others' work. Talk to a friend, go for a walk, meditate, switch to something else.
* Talk synchronously, e.g. call or in person, if there are too many "I didn't understand" or "Alternative solution:" comments. Prefer not to use chat for that. Post a follow-up
  comment summarizing the discussion.

> * Avoid selective ownership of code ("mine", "not mine", "yours"). Prefer referring to it as `our code`.

Reviewer 2: ```suggestion
* Avoid selective ownership of code ("mine", "not mine", "yours"). Prefer referring to it as `our` code.
```

> * Use "we" instead of "you", unless asking for opinion of specific individual ("What do you think ...?", but "Here we are doing ... What if we ...?",)

Reviewer 2: ```suggestion
* Use "we" instead of "you", unless asking for someon's opinion ("What do you think ...?", but "Here we are doing ... What if we ...?",)
```

* Avoid using terms that could be seen as referring to personal traits. ("dumb",
  "stupid", "insane"). Assume everyone is intelligent and well-meaning.
* Be aware of negative bias with online communication. (If content is neutral, we assume the tone is negative.) Can you use positive language as opposed to neutral?

> * When using emojis be aware that their meaning can be different for different individuals - winking smile may not be always innocent

Reviewer 2: ```suggestion
* When using emojis be aware that their meaning can be different for different individuals (e.g. a winking smile may not be always innocent)
```

* Same with any other media format you decide to use

> * Do not expect your comments to be answered immediately. If you need an answer fast - talk in person (the same applies to the Slack messages)

Reviewer 2: ```suggestion
* Do not expect your comments to be answered immediately. If you need an answer fast — talk directly with the person (the same applies to the Slack messages)
```

This is more because "in person" is a bit odd given that like half the team is remote

> * It's fine to have disagreements and it's hard to please everyone.

Author: ```suggestion
* Accept that it's fine to have disagreements and it's hard to please everyone
```

Creating a Pull Request
--------

> * When creating your branches, name them using a format like `<author>/<ticket>` or `<author>/<ticket>-<shortdescription>`

Reviewer 7: I wonder if this could be automated with `Steve` e.g when a person moves ticket in progress we could create branch if need with correct format. What do you guys think?

Author: you mean listen to JIRA webhooks? Interesting idea. But we can start with something simple like  a git hook maybe.

Reviewer 2: Nice idea @sergdort

* Carefully fill in the pull request template:
	* the title of the PR should follow the format `[XXX-123] Description of the change`, starting with a ticket number

> 	* explain the context and motivation, don’t assume familiarity with the history. Highlight how the goal was achieved without going too much into details - people tend not to read long descriptions and if the long explanation is needed, then just a single pull request is probably not gonna be enough (i.e. the change may need to go through the proposal process or design review). 

Reviewer 2: ```suggestion
	* explain the context and motivation, don’t assume familiarity with the history. Highlight how the goal was achieved without going too much into details - people tend not to read long descriptions and if a long explanation is needed, then a single pull request is probably not going to be enough (i.e. the change may need to go through the proposal process or design review). 
```

	* don't strip out parts of the template unless it is absolutely irrelevant to the changes (i.e. change in the build script does not need any screenshots as it does not touch UI)
	* don't forget labels (note: it's better to set them after the PR is created, as GitHub sometimes fails with 405 status code and we don't know why)
* Explicitly request review from or mention team mates you specifically want to involve in the discussion (in addition to the ones that will be assigned to it automatically by PullAssigners)
* Self-review your changes, adding comments where you think you can get questions (it may worth adding these as code or documentation comments). Tip: if you spot small issues you can use GitHub suggestions and apply them right away

> * Keep your pull requests small, ideally they shouldn't be more than **800 additions/deletions**

Author: Thanks, I'll update. btw I still don't see a reason to have "soft" and "hard" limit (850). It should be just one ("soft") and results in a warning by Danger.

* Create a draft PR for work that is in progress, but only do that if you seek for opinions on your work

> * Mark PR with the "Blocked" label if it has any dependency that should stop it from being merged (it's fine for pull request be both "blocked" and "ready for review", but it shouldn't be "ready for review" and "wip")

Reviewer 2: ```suggestion
* Mark a PR with the "Blocked" label if it has any dependency that should stop it from being merged (it's fine for pull requests to be both "blocked" and "ready for review", but it shouldn't be "ready for review" and "wip")
```

> * Mark a PR with the "Blocked" label if it has any dependency that should stop it from being merged (it's fine for pull requests to be both "blocked" and "ready for review", but it shouldn't be "ready for review" and "wip")

Reviewer 8: Shouldn't we say that if a PR is `blocked` it should mention the reason for it being blocked in the description? It can be blocked because another PR has to be merged first. In this case the reviewer can review the PR that is the blocker instead of that one.

Author: for sure, I'll add it

> * Give your team mates a reasonable time to review your changes, if the change is critical use "Top priority" label and post it in the team channel

Reviewer 2: ```suggestion
* Give your team members a reasonable time to review your changes, if the change is critical use "Top priority" label and post it in the team channel
```

I know it's stretching it but especially in the uk `mate` is not really gender neutral (learned this the other day…). Better safe than sorry :)

* Do not leave stale pull requests, seek for help/actions or close them (make sure you don't delete your local branch if you may need it later). If you find one that appears to have been stale for a week, please warn the author. After two weeks it can be closed with a note after a discussion of its current status with the author.
* It's our common effort to push product forward and we all do equally valuable work. Keep that in mind when creating a PR: try to review some of other's PRs, and avoid having more than X PRs opened at the same time

Reviewing a Pull Request
--------

> * Accept that many programming decisions are opinions. Discuss tradeoffs, which

Reviewer 2: ```suggestion
* Accept that many programming decisions are opinions. Discuss tradeoffs, which one
```

* Focus not on finding flaws in the code (no one writes perfect code) but on understanding the change in the first place
* Encourage things you like, try not only comment on what you don't like
* Before suggesting to change code think if it brings any real value or if it is just your personal preference. Is this code something you can live with or not?
* Do not try to make everything perfect (it never will be and otherwise we'll never deliver anything), but do not compromise on quality
* Do not comment if you don't have any suggestions to address your comment
* Use GitHub suggestions as much as possible
* If you feel strong against the change - request changes and explain why you are doing that. Do not request changes for typos, missing documentation comments or code style violations, trust your colleagues to address them.
* Approve changes if you are confident about them. 
* If you are the first reviewer to approve the changes add the label "Needs one reviewer" so that others can see which PRs already have a feedback and which don't have any yet.
* Comment on the changes if something is not clear to you and needs further clarification or actions

> * If you are not familiar with the domain ask for more context (better in person) rather than skipping to something more familiar

Reviewer 2: ```suggestion
* If you are not familiar with the domain ask for more context (better directly to the author) rather than skipping to something more familiar
```

Responding to feedback
-------------------------

* Wait before answering to a single comment, the reviewer may be still writing more of them (though they should prefer to send all comments at once)
* Be grateful for the reviewer's suggestions. ("Good call. I'll make that
  change.")
* If you disagree with a suggestion, explain why and avoid template answers like "out of the scope" or "thank you for your opinion". If you think that suggestion is a personal preference and does not bring value - say so
* Assume the best intention from the reviewer's comments.
* If a reviewer seems aggressive or angry or otherwise personal, consider if it is intended to be read that way and ask the person for clarification of intent, in person if possible. Do not escalate online discussions.
* If you can't reach agreement seek for more opinions
* Don't be too defensive about your code, someone will rewrite it anyway sooner or later (it might be yourself as well)

> * Try to answer all comments, don't make teammates feel like you are ignoring them. You can review your own code for that instead of sending individual replies so that all your answers are posted at once.

Reviewer 2: ```suggestion
* Try to answer all comments, don't make your colleagues feel like you are ignoring them. You can review your own code for that instead of sending individual replies so that all your answers are posted at once.
```

* Add "work in progress" label while addressing review comments
* Re-request review when you addressed the comments (either implementing suggested change or replying to the comment) and label it with "Ready for review" again
* Do not rebase and force-push to your pull request branch after you have received any comments, use merge instead. This allows reviewers to view changes on the PR since their last review, rather than re-reviewing a lot of the code they have already reviewed. All pull requests are squashed when they are merged into the base branch so the history of your changes won't matter. It's fine to rebase and force-push if there are no comments.

> * Do not dismiss any pending requests for changes unless you agreed with their authors that it was addressed

Reviewer 8: In this case shouldn't the author mention in a comment that the request has been discussed with the reviewer and the necessary action has been taken?

Author: yep, and github actually enforced to enter the reason for dismissal and adds it as a comment, will add this

Reviewer 8: Ahh great! Didn't know that 😄

* When you have two approvals and have no pending questions or requests for changes and don't wait for review from anyone else - put a "Merge" label for bot to merge your changes.

PR Parties
--------

To improve collaboration, we take time to review pull request together in small groups every week (on Monday). This is what we call "PR parties".

- Each group should include at least one senior developer. 
- Each group will have a few pull requests assigned to them
- When done with their assigned PRs, the group is free to review something else (i.e. each other's pull requests) or finish the meeting

Pull Assigners
--------

To balance the workload and avoid knowledge silos we use [Pull Assigners](https://pullpanda.com/assigner) to assign reviewers.

> - Each pull request will be assigned to three random reviewers from the team

Reviewer 2: ```suggestion
- Each pull request will be assigned to three random reviewers from the team (PS: if the PA bot seems stuck, try to de-assign it and re-assign it.)
```

> - Two approvals are enough to merge pull request (there shouldn't be pending requests for changes though)

Reviewer 2: ```suggestion
- Two approvals are enough to merge a pull request (there shouldn't be pending requests for changes though)
```

> - Anyone else is still encouraged to review any pull request and author can request reviews from specific teammates

Reviewer 8: What happens if a PR has two approvals but the author has requested review from a specific teammate that hasn't reviewed it yet? Shouldn't in this case the author wait for that specific review?

Author: yep, hence that's up to the author to put merge label (I think that's mentioned somewhere in the doc, will check that)

Reviewer 8: Yes, you are right, it is mentioned above. Leave it then 😊
